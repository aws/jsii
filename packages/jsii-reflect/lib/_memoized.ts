import { TypeSystem } from './type-system';

/* eslint-disable @typescript-eslint/ban-types -- WeakMap<T, _> demands T extends object */
const CACHE = new WeakMap<object, Map<string, unknown>>();

function memoizedGet(original: () => any, propertyKey: string): () => any {
  return function (this: object): unknown {
    let cache = CACHE.get(this);
    if (cache == null) {
      cache = new Map();
      CACHE.set(this, cache);
    }
    if (cache.has(propertyKey)) {
      const result = cache.get(propertyKey);
      if (Array.isArray(result)) {
        // Return a copy of arrays as a precaution
        return Array.from(result);
      }
      return result;
    }
    const result = original.call(this);
    // If the result is an array, memoize a copy for safety.
    cache.set(propertyKey, Array.isArray(result) ? Array.from(result) : result);
    return result;
  };
}

/**
 * Decorates property readers for readonly properties so that their results are
 * memoized in a `WeakMap`-based cache. Those properties will consequently be
 * computed exactly once.
 *
 * This can only be applied to property accessors (`public get foo(): any`), and not to
 * property declarations (`public readonly foo: any`).
 *
 * This should not be applied to any computations relying on a typesystem.
 * The typesystem can be changed and thus change the result of the call.
 * Use `memoizedWhenLocked` instead.
 */
export function memoized(
  _prototype: unknown,
  propertyKey: string,
  descriptor: PropertyDescriptor,
): void {
  if (!descriptor.get) {
    throw new Error(`@memoized can only be applied to property getters!`);
  }
  if (descriptor.set) {
    throw new Error(`@memoized can only be applied to readonly properties!`);
  }

  const original = descriptor.get;
  descriptor.get = memoizedGet(original, propertyKey);
}

export function memoizedWhenLocked<T extends { system: TypeSystem }>(
  _prototype: T,
  propertyKey: string,
  descriptor: PropertyDescriptor,
): void {
  if (!descriptor.get) {
    throw new Error(`@memoized can only be applied to property getters!`);
  }
  if (descriptor.set) {
    throw new Error(`@memoized can only be applied to readonly properties!`);
  }

  const original = descriptor.get;
  descriptor.get = function (this: T): unknown {
    if (this.system.isLocked) {
      return memoizedGet(original, propertyKey).call(this);
    }

    return original.call(this);
  };
}
