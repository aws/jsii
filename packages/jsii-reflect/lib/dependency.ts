import { TypeSystem } from './type-system';

export class Dependency {
  public constructor(
    public readonly system: TypeSystem,
    private readonly name: string,
    public readonly version: string,
  ) {}

  public get assembly() {
    return this.system.findAssembly(this.name);
  }
}
