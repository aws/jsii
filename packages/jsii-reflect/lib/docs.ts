import * as jsii from '@jsii/spec';
import { Stability } from '@jsii/spec';
import { TypeSystem } from './type-system';

export interface Documentable {
  docs: Docs;
}

export class Docs {
  public readonly docs: jsii.Docs;

  public constructor(
    public readonly system: TypeSystem,
    public readonly target: Documentable,
    spec: jsii.Docs,
    private readonly parentDocs?: Docs,
  ) {
    this.docs = spec ?? {};
  }

  public toString() {
    const ret = new Array<string>();
    if (this.docs.summary) {
      ret.push(this.docs.summary);
    }
    if (this.docs.remarks) {
      ret.push('', this.docs.remarks);
    }
    return ret.join('\n');
  }

  public get subclassable(): boolean {
    return !!this.docs.subclassable;
  }

  /**
   * Return the reason for deprecation of this type
   */
  public get deprecationReason(): string | undefined {
    if (this.docs.deprecated !== undefined) {
      return this.docs.deprecated;
    }
    if (this.parentDocs) {
      return this.parentDocs.deprecationReason;
    }
    return undefined;
  }

  /**
   * Return whether this type is deprecated
   */
  public get deprecated(): boolean {
    return this.deprecationReason !== undefined;
  }

  /**
   * Return the stability of this type
   */
  public get stability(): jsii.Stability | undefined {
    return lowestStability(this.docs.stability, this.parentDocs?.stability);
  }

  public customTag(tag: string): string | undefined {
    return this.docs.custom?.[tag];
  }

  public get summary(): string {
    return this.docs.summary ?? '';
  }

  public get remarks(): string {
    return this.docs.remarks ?? '';
  }
}

const stabilityPrecedence = {
  [Stability.Deprecated]: 0,
  [Stability.Experimental]: 1,
  [Stability.External]: 2,
  [Stability.Stable]: 3,
};

function lowestStability(a?: Stability, b?: Stability): Stability | undefined {
  if (a === undefined) {
    return b;
  }
  if (b === undefined) {
    return a;
  }
  return stabilityPrecedence[a] < stabilityPrecedence[b] ? a : b;
}
