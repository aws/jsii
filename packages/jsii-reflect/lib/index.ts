export * from './assembly';
export * from './class';
export * from './dependency';
export * from './docs';
export * from './enum';
export * from './interface';
export * from './method';
export * from './overridable';
export * from './parameter';
export * from './property';
export * from './tree';
export * from './type-ref';
export * from './type-system';
export * from './type';
export * from './reference-type';
export * from './type-member';
