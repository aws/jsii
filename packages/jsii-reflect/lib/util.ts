
export function indexBy<T>(xs: T[], f: (x: T) => string): {[key: string]: T} {
  const ret: {[key: string]: T} = {};
  for (const x of xs) {
    ret[f(x)] = x;
  }
  return ret;
}