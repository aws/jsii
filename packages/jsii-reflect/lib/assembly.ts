import jsii = require('jsii-spec');
import { ClassType } from './class';
import { Dependency } from './dependency';
import { EnumType } from './enum';
import { InterfaceType } from './interface';
import { Type } from './type';
import { TypeSystem } from './type-system';

export class Assembly {
  private _typeCache?: { [fqn: string]: Type };
  private _dependencyCache?: { [name: string]: Dependency };

  constructor(
    public readonly system: TypeSystem,
    private readonly spec: jsii.Assembly) { }

  /**
   * The version of the spec schema
   */
  public get schema(): jsii.SchemaVersion {
    return this.spec.schema;
  }

  /**
   * The name of the assembly
   */
  public get name(): string {
    return this.spec.name;
  }

  /**
   * Description of the assembly, maps to "description" from package.json
   * This is required since some package managers (like Maven) require it.
   */
  public get description(): string {
    return this.spec.description;
  }

  /**
   * The url to the project homepage. Maps to "homepage" from package.json.
   */
  public get homepage(): string {
    return this.spec.homepage;
  }

  /**
   * The module repository, maps to "repository" from package.json
   * This is required since some package managers (like Maven) require it.
   */
  public get repository(): { type: string, url: string } {
    return this.spec.repository;
  }

  /**
   * The main author of this package.
   */
  public get author(): jsii.Person {
    return this.spec.author;
  }

  /**
   * Additional contributors to this package.
   */
  public get contributors(): jsii.Person[] {
    return this.spec.contributors || [ ];
  }

  /**
   * A fingerprint that can be used to determine if the specification has changed.
   */
  public get fingerprint(): string {
    return this.spec.fingerprint;
  }

  /**
   * The version of the assembly
   */
  public get version(): string {
    return this.spec.version;
  }

  /**
   * The SPDX name of the license this assembly is distributed on.
   */
  public get license(): string {
    return this.spec.license;
  }

  /**
   * A map of target name to configuration, which is used when generating packages for
   * various languages.
   */
  public get targets() {
    return this.spec.targets;
  }

  /**
   * Dependencies on other assemblies (with semver), the key is the JSII assembly name.
   */
  public get dependencies(): Dependency[] {
    return Object.keys(this._dependencies).map(name => this._dependencies[name]);
  }

  public findDependency(name: string) {
    const dep = this._dependencies[name];
    if (!dep) {
      throw new Error(`Dependency ${name} not found for assembly ${this.name}`);
    }
    return dep;
  }

  /**
   * List if bundled dependencies (these are not expected to be jsii assemblies).
   */
  public get bundled(): { [module: string]: string } {
    return this.spec.bundled || { };
  }

  /**
   * The top-level readme document for this assembly (if any).
   */
  public get readme() {
    return this.spec.readme;
  }

  /**
   * All types in the assembly, keyed by their fully-qualified-name
   */
  public get types(): Type[] {
    return Object.keys(this._types).map(key => this._types[key]);
  }

  public get classes(): ClassType[] {
    return this.types.filter(t => t instanceof ClassType).map(t => t as ClassType);
  }

  public get interfaces(): InterfaceType[] {
    return this.types.filter(t => t instanceof InterfaceType).map(t => t as InterfaceType);
  }

  public get enums(): EnumType[] {
    return this.types.filter(t => t instanceof EnumType).map(t => t as EnumType);
  }

  public findType(fqn: string) {
    const type = this._types[fqn];
    if (!type) {
      throw new Error(`Type '${fqn}' not found in assembly ${this.name}`);
    }
    return type;
  }

  public tryFindType(fqn: string): Type | undefined {
    return this._types[fqn];
  }

  private get _dependencies() {
    if (!this._dependencyCache) {
      this._dependencyCache = { };
      if (this.spec.dependencies) {
        for (const name of Object.keys(this.spec.dependencies)) {
          this._dependencyCache[name] = new Dependency(this.system, name, this.spec.dependencies[name]);
        }
      }
    }

    return this._dependencyCache;
  }

  private get _types() {
    if (!this._typeCache) {
      this._typeCache = { };

      const ts = this.spec.types || { };
      for (const fqn of Object.keys(ts)) {
        const type = ts[fqn];
        switch (type.kind) {
          case jsii.TypeKind.Class:
            this._typeCache[fqn] = new ClassType(this.system, this, type);
            break;

          case jsii.TypeKind.Interface:
            this._typeCache[fqn] = new InterfaceType(this.system, this, type);
            break;

          case jsii.TypeKind.Enum:
            this._typeCache[fqn] = new EnumType(this.system, this, type);
            break;

          default:
            throw new Error(`Unknown type kind`);
        }
      }
    }

    return this._typeCache;
  }
}
