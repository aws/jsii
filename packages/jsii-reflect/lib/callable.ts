import * as jsii from '@jsii/spec';
import { Assembly } from './assembly';
import { Docs, Documentable } from './docs';
import { Overridable } from './overridable';
import { Parameter } from './parameter';
import { SourceLocatable } from './source';
import { locationInRepository, SourceLocation } from './source';
import { Type } from './type';
import { MemberKind, TypeMember } from './type-member';
import { TypeSystem } from './type-system';

export abstract class Callable
  implements Documentable, Overridable, TypeMember, SourceLocatable {
  public abstract readonly kind: MemberKind;
  public abstract readonly name: string;
  public abstract readonly abstract: boolean;

  public constructor(
    public readonly system: TypeSystem,
    public readonly assembly: Assembly,
    public readonly parentType: Type,
    public readonly spec: jsii.Callable,
  ) {}

  /**
   * The parameters of the method/initializer
   */
  public get parameters(): Parameter[] {
    return (this.spec.parameters ?? []).map(
      (p) => new Parameter(this.system, this.parentType, this, p),
    );
  }

  /**
   * Indicates if this method is protected (otherwise it is public)
   */
  public get protected(): boolean {
    return !!this.spec.protected;
  }

  /**
   * Indicates whether this method is variadic or not. When ``true``, the last
   * element of ``#parameters`` will also be flagged ``#variadic``.
   */
  public get variadic(): boolean {
    return !!this.spec.variadic;
  }

  public get overrides(): Type | undefined {
    if (!this.spec.overrides) {
      return undefined;
    }

    return this.system.findFqn(this.spec.overrides);
  }

  public get docs(): Docs {
    return new Docs(
      this.system,
      this,
      this.spec.docs ?? {},
      this.parentType.docs,
    );
  }

  /**
   * Return the location in the module
   */
  public get locationInModule(): SourceLocation | undefined {
    return this.spec.locationInModule;
  }

  /**
   * Return the location in the repository
   */
  public get locationInRepository(): SourceLocation | undefined {
    return locationInRepository(this);
  }

  public toString() {
    return `${this.kind}:${this.parentType.fqn}.${this.name}`;
  }
}
