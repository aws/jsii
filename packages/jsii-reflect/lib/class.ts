import jsii = require('jsii-spec');
import { Assembly } from './assembly';
import { InterfaceType } from './interface';
import { Method } from './method';
import { Property } from './property';
import { ReferenceType } from './reference-type';
import { TypeSystem } from './type-system';
import { indexBy } from './util';

export class ClassType extends ReferenceType {
  constructor(
    public readonly system: TypeSystem,
    public readonly assembly: Assembly,
    private readonly classSpec: jsii.ClassType) {
    super(system, assembly, classSpec);
  }

  /**
   * Base class (optional).
   */
  public get base(): ClassType | undefined {
    if (!this.classSpec.base) {
      return undefined;
    }

    const type = this.system.findFqn(this.classSpec.base.fqn);
    if (!(type instanceof ClassType)) {
      throw new Error(`FQN for base class points to a non-class type: ${this.classSpec.base.fqn}`);
    }

    return type;
  }

  /**
   * Initializer (constructor) method.
   */
  public get initializer(): Method | undefined {
    if (!this.classSpec.initializer) {
      return undefined;
    }

    return new Method(this.system, this.assembly, this, this.classSpec.initializer);
  }

  /**
   * Indicates if this class is an abstract class.
   */
  public get abstract(): boolean {
    return !!this.classSpec.abstract;
  }

  /**
   * Returns list of all base classes (first is the direct base and last is the top-most).
   */
  public getAncestors() {
    const out = new Array<ClassType>();
    if (this.base) {
      out.push(this.base);
      out.push(...this.base.getAncestors());
    }
    return out;
  }

  /**
   * Lists all properties in this class.
   * @param inherited include all properties inherited from base classes (default: false)
   */
  public getProperties(inherited = false): {[name: string]: Property} {
    const base = inherited && this.base ? this.base.getProperties(inherited) : {};
    return Object.assign(base, indexBy(
      (this.classSpec.properties || []).map(p => new Property(this.system, this.assembly, this, p)),
      p => p.name));
  }

  /**
   * List all methods in this class.
   * @param inherited include all methods inherited from base classes (default: false)
   */
  public getMethods(inherited = false): {[name: string]: Method} {
    const base = inherited && this.base ? this.base.getMethods(inherited) : {};
    return Object.assign(base, indexBy(
      (this.classSpec.methods || []).map(m => new Method(this.system, this.assembly, this, m)),
      m => m.name));
  }

  /**
   * Lists all interfaces this class implements.
   * @param inherited include all interfaces implemented by all base classes (default: false)
   */
  public getInterfaces(inherited = false) {
    const out = new Array<InterfaceType>();
    if (inherited && this.base) {
      out.push(...this.base.getInterfaces(inherited));
    }
    if (this.classSpec.interfaces) {
      out.push(...this.classSpec.interfaces.map(i => this.system.findInterface(i.fqn)));
    }
    return out;
  }

  public isClassType() {
    return true;
  }
}
