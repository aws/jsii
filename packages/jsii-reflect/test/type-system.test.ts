import * as spec from '@jsii/spec';
import { Stability } from '@jsii/spec';
import * as path from 'path';
import { TypeSystem } from '../lib';
import { typeSystemFromSource } from './util';

let typesys: TypeSystem;

beforeAll(async () => {
  typesys = new TypeSystem();
  await typesys.loadModule(resolveModuleDir('jsii-calc'));
});

test('TypeSystem.hasAssembly', () => {
  expect(typesys.includesAssembly('@foo/bar')).toBeFalsy();
  expect(typesys.includesAssembly('jsii-calc')).toBeTruthy();
  expect(typesys.includesAssembly('@scope/jsii-calc-lib')).toBeTruthy();
});

test('TypeSystem.assemblies lists all the loaded assemblies', () =>
  expect(typesys.assemblies.map((a) => a.name).sort()).toMatchSnapshot());

test('TypeSystem.classes lists all the classes in the typesystem', () =>
  expect(typesys.classes.map((c) => c.fqn).sort()).toMatchSnapshot());

test('findClass', () => {
  const calc = typesys.findClass('jsii-calc.Calculator');
  const actual = new Array<string>();
  Object.values(calc.getMethods(/* inherited */ true)).forEach((method) => {
    actual.push(`${method.name} from ${method.parentType.name}`);
  });

  expect(actual.join('\n')).toMatchSnapshot();
});

test('"roots" is a list of the directly loaded assemblies', async () => {
  expect(typesys.roots.length).toBe(1);
  expect(typesys.roots[0]).toBe(typesys.findAssembly('jsii-calc'));

  // now load another assembliy directly
  await typesys.load(resolveModuleDir('@scope/jsii-calc-lib'));
  return expect(typesys.roots.length).toBe(2);
});

describe('Type', () => {
  test('.isClassType', () => {
    // GIVEN
    const clazz = typesys.findFqn('jsii-calc.AllTypes');
    const iface = typesys.findFqn('jsii-calc.IPublicInterface');
    const enumt = typesys.findFqn('jsii-calc.AllTypesEnum');

    // THEN
    expect(clazz.isClassType()).toBeTruthy();
    expect(iface.isClassType()).toBeFalsy();
    expect(enumt.isClassType()).toBeFalsy();
  });

  test('.isDataType', () => {
    // GIVEN
    const clazz = typesys.findFqn('jsii-calc.AllTypes');
    const iface = typesys.findFqn(
      'jsii-calc.IInterfaceThatShouldNotBeADataType',
    );
    const datat = typesys.findFqn('jsii-calc.CalculatorProps');
    const enumt = typesys.findFqn('jsii-calc.AllTypesEnum');

    // THEN
    expect(clazz.isDataType()).toBeFalsy();
    expect(iface.isDataType()).toBeFalsy();
    expect(datat.isDataType()).toBeTruthy();
    expect(enumt.isDataType()).toBeFalsy();
  });

  test('.isInterfaceType', () => {
    // GIVEN
    const clazz = typesys.findFqn('jsii-calc.AllTypes');
    const iface = typesys.findFqn('jsii-calc.IPublicInterface');
    const enumt = typesys.findFqn('jsii-calc.AllTypesEnum');

    // THEN
    expect(clazz.isInterfaceType()).toBeFalsy();
    expect(iface.isInterfaceType()).toBeTruthy();
    expect(enumt.isInterfaceType()).toBeFalsy();
  });

  test('.isEnumType', () => {
    // GIVEN
    const clazz = typesys.findFqn('jsii-calc.AllTypes');
    const iface = typesys.findFqn('jsii-calc.IPublicInterface');
    const enumt = typesys.findFqn('jsii-calc.AllTypesEnum');

    // THEN
    expect(clazz.isEnumType()).toBeFalsy();
    expect(iface.isEnumType()).toBeFalsy();
    expect(enumt.isEnumType()).toBeTruthy();
  });

  describe('.extends(base)', () => {
    test('with interfaces', () => {
      // GIVEN
      const base = typesys.findFqn(
        '@scope/jsii-calc-base-of-base.VeryBaseProps',
      );
      const clazz = typesys.findFqn('jsii-calc.ImplictBaseOfBase');
      const enumt = typesys.findFqn('jsii-calc.AllTypesEnum');

      // THEN
      expect(base.extends(base)).toBeTruthy();
      expect(clazz.extends(clazz)).toBeTruthy();
      expect(clazz.extends(base)).toBeTruthy();
      expect(enumt.extends(base)).toBeFalsy();
      expect(base.extends(enumt)).toBeFalsy();
      expect(base.extends(clazz)).toBeFalsy();
    });

    test('with a class and an interface', () => {
      // GIVEN
      const iface = typesys.findFqn(
        'jsii-calc.IInterfaceImplementedByAbstractClass',
      );
      const clazz = typesys.findFqn('jsii-calc.AbstractClass');

      // THEN
      expect(clazz.extends(iface)).toBeTruthy();
    });

    test('with two classes', () => {
      // GIVEN
      const base = typesys.findFqn('jsii-calc.AbstractClassBase');
      const clazz = typesys.findFqn('jsii-calc.AbstractClass');

      // THEN
      expect(clazz.extends(base)).toBeTruthy();
    });
  });

  describe('.allImplementations', () => {
    test('with an interface', () => {
      // GIVEN
      const base = typesys.findFqn(
        'jsii-calc.IInterfaceImplementedByAbstractClass',
      );

      // THEN
      expect(base.allImplementations).toEqual([
        typesys.findFqn('jsii-calc.AbstractClass'),
        base,
      ]);
    });

    test('with an enum', () => {
      // GIVEN
      const enumt = typesys.findFqn('jsii-calc.AllTypesEnum');

      // THEN
      expect(enumt.allImplementations).toEqual([]);
    });
  });
});

test('Three Inheritance Levels', () => {
  const iface = typesys.findInterface(
    '@scope/jsii-calc-lib.IThreeLevelsInterface',
  );
  const methodnames = iface.allMethods.map((m) => m.name);
  methodnames.sort();
  expect(methodnames).toEqual(['bar', 'baz', 'foo']);
});

describe('@deprecated', () => {
  test('can be read on an item', () => {
    const klass = typesys.findClass('jsii-calc.Old');
    expect(klass.docs.deprecated).toBeTruthy();
    expect(klass.docs.stability).toBe(spec.Stability.Deprecated);
  });

  test('is inherited from class', () => {
    const klass = typesys.findClass('jsii-calc.Old');
    const method = klass.getMethods().doAThing;
    expect(method.docs.deprecated).toBeTruthy();
  });
});

test('overridden member knows about both parent types', async () => {
  const ts = await typeSystemFromSource(`
    export class Foo {
      public bar() {
        Array.isArray(3);
      }

      public boo(): boolean {
        return true;
      }
    }

    export class SubFoo extends Foo {
      public boo(): boolean {
        return false;
      }
    }
  `);

  const superType = ts.findClass('testpkg.Foo');
  const subType = ts.findClass('testpkg.SubFoo');
  const fooMethod = subType.allMethods.find((m) => m.name === 'bar')!;
  const booMethod = subType.allMethods.find((m) => m.name === 'boo')!;

  expect(fooMethod.parentType).toBe(subType);
  expect(fooMethod.definingType).toBe(superType);

  expect(booMethod.parentType).toBe(subType);
  expect(booMethod.definingType).toBe(subType);
  expect(booMethod.overrides).toBe(superType);
});

describe('Stability', () => {
  test('can be read on an item', () => {
    const klass = typesys.findClass('jsii-calc.DocumentedClass');
    expect(klass.docs.stability).toBe(spec.Stability.Stable);
  });

  test('is inherited from class', () => {
    const klass = typesys.findClass('jsii-calc.DocumentedClass');
    const method = klass.getMethods().greet;
    expect(method.docs.stability).toBe(spec.Stability.Stable);
  });

  test('can be overridden from class', () => {
    const klass = typesys.findClass('jsii-calc.DocumentedClass');
    const method = klass.getMethods().hola;
    expect(method.docs.stability).toBe(spec.Stability.Experimental);
  });

  // ----------------------------------------------------------------------

  describe('lowest stability guarantee is advertised', () => {
    test('when subclass is experimental', async () => {
      const ts = await typeSystemFromSource(`
        /**
         * @stable
         */
        export class Foo {
          constructor() {
            Array.isArray(3);
          }

          public bar() {
            Array.isArray(3);
          }
        }

        /**
         * @experimental
         */
        export class SubFoo extends Foo {
        }
      `);
      const classType = ts.findClass('testpkg.SubFoo');
      const initializer = classType.initializer!;
      const method = classType.allMethods.find((m) => m.name === 'bar')!;

      expect(initializer.docs.stability).toEqual(Stability.Experimental);
      expect(method.docs.stability).toEqual(Stability.Experimental);
    });

    test('when method is experimental', async () => {
      const ts = await typeSystemFromSource(`
        /**
         * @stable
         */
        export class Foo {
          /**
           * @experimental
           */
          constructor() {
            Array.isArray(3);
          }

          /**
           * @experimental
           */
          public bar() {
            Array.isArray(3);
          }
        }

        /**
         * @stable
         */
        export class SubFoo extends Foo {
        }
      `);

      const classType = ts.findClass('testpkg.SubFoo');
      const initializer = classType.initializer!;
      const method = classType.allMethods.find((m) => m.name === 'bar')!;

      expect(initializer.docs.stability).toEqual(Stability.Experimental);
      expect(method.docs.stability).toEqual(Stability.Experimental);
    });

    test('when method is explicitly marked stable', async () => {
      const ts = await typeSystemFromSource(`
        /**
         * @stable
         */
        export class Foo {
          /**
           * @stable
           */
          constructor() {
            Array.isArray(3);
          }

          /**
           * @stable
           */
          public bar() {
            Array.isArray(3);
          }
        }

        /**
         * @experimental
         */
        export class SubFoo extends Foo {
        }
      `);

      const classType = ts.findClass('testpkg.SubFoo');
      const initializer = classType.initializer!;
      const method = classType.allMethods.find((m) => m.name === 'bar')!;

      expect(initializer.docs.stability).toEqual(Stability.Experimental);
      expect(method.docs.stability).toEqual(Stability.Experimental);
    });

    test('external stability', async () => {
      const ts = await typeSystemFromSource(`
        /**
         * @stability external
         */
        export class Foo {
          public bar() {
            Array.isArray(3);
          }
        }

        /**
         * @stable
         */
        export class SubFoo extends Foo {
        }
      `);

      const classType = ts.findClass('testpkg.SubFoo');
      const method = classType.allMethods.find((m) => m.name === 'bar')!;

      expect(method.docs.stability).toEqual(Stability.External);
    });
  });
});

test('TypeSystem.properties', async () => {
  const ts = await typeSystemFromSource(`
  export namespace submodule {
    export class Foo {
      public readonly test = 'TEST';
    }
    export interface Bar {
      readonly baz: number;
    }
  }
  `);
  expect(ts.properties).toHaveLength(2);
});

test('TypeSystem.methods', async () => {
  const ts = await typeSystemFromSource(`
  export namespace submodule {
    export class Foo {
      public method(): void {}
    }
    export interface IBar {
      baz(): number;
    }
  }
  `);
  expect(ts.methods).toHaveLength(2);
});

function resolveModuleDir(name: string) {
  return path.dirname(require.resolve(`${name}/package.json`));
}
