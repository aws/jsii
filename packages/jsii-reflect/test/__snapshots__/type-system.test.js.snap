// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeSystem.assemblies lists all the loaded assemblies 1`] = `
Array [
  "@scope/jsii-calc-base",
  "@scope/jsii-calc-base-of-base",
  "@scope/jsii-calc-lib",
  "jsii-calc",
]
`;

exports[`TypeSystem.classes lists all the classes in the typesystem 1`] = `
Array [
  "@scope/jsii-calc-base-of-base.Very",
  "@scope/jsii-calc-base.Base",
  "@scope/jsii-calc-lib.Number",
  "@scope/jsii-calc-lib.Operation",
  "@scope/jsii-calc-lib.Value",
  "@scope/jsii-calc-lib.submodule.Reflector",
  "jsii-calc.AbstractClass",
  "jsii-calc.AbstractClassBase",
  "jsii-calc.AbstractClassReturner",
  "jsii-calc.AbstractSuite",
  "jsii-calc.Add",
  "jsii-calc.AllTypes",
  "jsii-calc.AllowedMethodNames",
  "jsii-calc.AmbiguousParameters",
  "jsii-calc.AnonymousImplementationProvider",
  "jsii-calc.AsyncVirtualMethods",
  "jsii-calc.AugmentableClass",
  "jsii-calc.BaseJsii976",
  "jsii-calc.Bell",
  "jsii-calc.BinaryOperation",
  "jsii-calc.Calculator",
  "jsii-calc.ClassThatImplementsTheInternalInterface",
  "jsii-calc.ClassThatImplementsThePrivateInterface",
  "jsii-calc.ClassWithCollections",
  "jsii-calc.ClassWithDocs",
  "jsii-calc.ClassWithJavaReservedWords",
  "jsii-calc.ClassWithMutableObjectLiteralProperty",
  "jsii-calc.ClassWithPrivateConstructorAndAutomaticProperties",
  "jsii-calc.ConfusingToJackson",
  "jsii-calc.ConstructorPassesThisOut",
  "jsii-calc.Constructors",
  "jsii-calc.ConsumePureInterface",
  "jsii-calc.ConsumerCanRingBell",
  "jsii-calc.ConsumersOfThisCrazyTypeSystem",
  "jsii-calc.DataRenderer",
  "jsii-calc.DefaultedConstructorArgument",
  "jsii-calc.Demonstrate982",
  "jsii-calc.DeprecatedClass",
  "jsii-calc.DerivedClassHasNoProperties.Base",
  "jsii-calc.DerivedClassHasNoProperties.Derived",
  "jsii-calc.DisappointingCollectionSource",
  "jsii-calc.DoNotOverridePrivates",
  "jsii-calc.DoNotRecognizeAnyAsOptional",
  "jsii-calc.DocumentedClass",
  "jsii-calc.DontComplainAboutVariadicAfterOptional",
  "jsii-calc.DoubleTrouble",
  "jsii-calc.EnumDispenser",
  "jsii-calc.EraseUndefinedHashValues",
  "jsii-calc.ExperimentalClass",
  "jsii-calc.ExportedBaseClass",
  "jsii-calc.ExternalClass",
  "jsii-calc.GiveMeStructs",
  "jsii-calc.GreetingAugmenter",
  "jsii-calc.ImplementInternalInterface",
  "jsii-calc.Implementation",
  "jsii-calc.ImplementsInterfaceWithInternal",
  "jsii-calc.ImplementsInterfaceWithInternalSubclass",
  "jsii-calc.ImplementsPrivateInterface",
  "jsii-calc.InbetweenClass",
  "jsii-calc.InterfaceCollections",
  "jsii-calc.InterfaceInNamespaceIncludesClasses.Foo",
  "jsii-calc.InterfacesMaker",
  "jsii-calc.JSII417Derived",
  "jsii-calc.JSII417PublicBaseOfBase",
  "jsii-calc.JSObjectLiteralForInterface",
  "jsii-calc.JSObjectLiteralToNative",
  "jsii-calc.JSObjectLiteralToNativeClass",
  "jsii-calc.JavaReservedWords",
  "jsii-calc.Jsii487Derived",
  "jsii-calc.Jsii496Derived",
  "jsii-calc.JsiiAgent",
  "jsii-calc.JsonFormatter",
  "jsii-calc.MethodNamedProperty",
  "jsii-calc.Multiply",
  "jsii-calc.Negate",
  "jsii-calc.NodeStandardLibrary",
  "jsii-calc.NullShouldBeTreatedAsUndefined",
  "jsii-calc.NumberGenerator",
  "jsii-calc.ObjectRefsInCollections",
  "jsii-calc.ObjectWithPropertyProvider",
  "jsii-calc.Old",
  "jsii-calc.OptionalArgumentInvoker",
  "jsii-calc.OptionalConstructorArgument",
  "jsii-calc.OptionalStructConsumer",
  "jsii-calc.OverridableProtectedMember",
  "jsii-calc.OverrideReturnsObject",
  "jsii-calc.PartiallyInitializedThisConsumer",
  "jsii-calc.Polymorphism",
  "jsii-calc.Power",
  "jsii-calc.PropertyNamedProperty",
  "jsii-calc.PublicClass",
  "jsii-calc.PythonReservedWords",
  "jsii-calc.ReferenceEnumFromScopedPackage",
  "jsii-calc.ReturnsPrivateImplementationOfInterface",
  "jsii-calc.RootStructValidator",
  "jsii-calc.RuntimeTypeChecking",
  "jsii-calc.SingleInstanceTwoTypes",
  "jsii-calc.SingletonInt",
  "jsii-calc.SingletonString",
  "jsii-calc.SomeTypeJsii976",
  "jsii-calc.StableClass",
  "jsii-calc.StaticContext",
  "jsii-calc.Statics",
  "jsii-calc.StripInternal",
  "jsii-calc.StructPassing",
  "jsii-calc.StructUnionConsumer",
  "jsii-calc.Sum",
  "jsii-calc.SupportsNiceJavaBuilder",
  "jsii-calc.SupportsNiceJavaBuilderWithRequiredProps",
  "jsii-calc.SyncVirtualMethods",
  "jsii-calc.Thrower",
  "jsii-calc.UnaryOperation",
  "jsii-calc.UpcasingReflectable",
  "jsii-calc.UseBundledDependency",
  "jsii-calc.UseCalcBase",
  "jsii-calc.UsesInterfaceWithProperties",
  "jsii-calc.VariadicInvoker",
  "jsii-calc.VariadicMethod",
  "jsii-calc.VirtualMethodPlayground",
  "jsii-calc.VoidCallback",
  "jsii-calc.WithPrivatePropertyInConstructor",
  "jsii-calc.composition.CompositeOperation",
  "jsii-calc.submodule.MyClass",
  "jsii-calc.submodule.nested_submodule.Namespaced",
]
`;

exports[`findClass 1`] = `
"typeName from Calculator
toString from Calculator
add from Calculator
mul from Calculator
neg from Calculator
pow from Calculator
readUnionValue from Calculator"
`;
